### ISSUE TITLE
Llama model not implemented

### ISSUE THREAD

**Issue by gerald4** [UFO-101/auto-circuit#13]
Hi!
I'd like to know how to use Llama models with auto-circuits.
My naive way has been to create a patchable model with 

model = patchable_model(
    hf_model,
    factorized=True,
    slice_output="last_seq",
    separate_qkv=True,
)
where hf_model is from the LlamaForCausalLM class. However, this does not work.
Thanks!

**Comment by UFO-101** [comment]
Hi! auto-circuit currently only supports transformer lens models. So you need to load the HookedTransformer from [TransformerLens](https://github.com/TransformerLensOrg/TransformerLens) and then pass it to patchable_model. However, auto-circuit is very memory intensive, so it may not be possible to load models with more than 1 billion parameters (unless `factorized=False`).

```
import transformer_lens

# Load a model (eg GPT-2 Small)
model = transformer_lens.HookedTransformer.from_pretrained("gpt2-small")
```

**Comment by gerald4** [comment]
There have been big discrepancies between the Llama models wrapped to HookedTransformers and Llama from HuggingFace: https://github.com/TransformerLensOrg/TransformerLens/issues/615

I am curious if your patch model class cannot be generalized so a to directly load Hugging Face models or models from other patching libraries such as nnsight https://github.com/ndif-team/nnsight.

**Comment by UFO-101** [comment]
> I am curious if your patch model class cannot be generalized so a to directly load Hugging Face models or models from other patching libraries such as nnsight https://github.com/ndif-team/nnsight.

It could be! You would need to add a file similar to [transformer_lens_utils.py](https://github.com/UFO-101/auto-circuit/blob/main/auto_circuit/model_utils/transformer_lens_utils.py) and then add an if statement in [this](https://github.com/UFO-101/auto-circuit/blob/dccddb557376eee22e5e320e8a1cc2503d026136/auto_circuit/utils/graph_utils.py#L166) function.

I'd welcome PRs to do this.

**Comment by gerald4** [comment]
Am I correct if I say that I need to only implement the factorized_src_nodes and factorized_dest_nodes functions?

**Comment by UFO-101** [comment]
Yes if you always want to use it as a factorized graph then that would be sufficient.

**Comment by gerald4** [comment]
Thank you! I'm working on that. Do you have other suggestions on how to integrate without using a factorized graph?
I'm just trying to know the easiest way compatible to your library.

**Comment by UFO-101** [comment]
If you want to do [edge patching](https://www.lesswrong.com/posts/caZ3yR5GnzbZe2yJ3/how-to-do-patching-fast#Edge_Patching) you need to use a factorized graph. If you instead want to do [node patching](https://www.lesswrong.com/posts/caZ3yR5GnzbZe2yJ3/how-to-do-patching-fast#Node_Patching),
I would suggest not using my library and instead using TransformerLens
hooks.
